README.rst
=========
pylexique
=========


.. image:: https://img.shields.io/badge/Maintained%3F-yes-green.svg
        :target: https://GitHub.com/SekouDiaoNlp/pylexique/graphs/commit-activity
        :alt: Package Maintenance Status

.. image:: https://img.shields.io/badge/maintainer-SekouDiaoNlp-blue
        :target: https://GitHub.com/SekouDiaoNlp/pylexique
        :alt: Package Maintener

.. image:: https://img.shields.io/github/checks-status/SekouDiaoNlp/mlconjug3/master?label=Build%20status%20on%20Windows%2C%20MacOs%20and%20Linux
        :target: https://github.com/SekouDiaoNlp/pylexique/actions/workflows/main.yml
        :alt: Build status on Windows, MacOs and Linux

.. image:: https://img.shields.io/pypi/v/pylexique.svg
        :target: https://pypi.python.org/pypi/pylexique
        :alt: Pypi Python Package Index Status

.. image:: https://anaconda.org/conda-forge/pylexique/badges/version.svg
        :target: https://anaconda.org/conda-forge/pylexique
        :alt: Anaconda Package Index Status

.. image:: https://img.shields.io/pypi/pyversions/pylexique
        :target: https://pypi.python.org/pypi/pylexique
        :alt: Compatible Python versions

.. image:: https://img.shields.io/conda/pn/conda-forge/pylexique?color=dark%20green&label=Supported%20platforms
        :target: https://anaconda.org/conda-forge/pylexique
        :alt: Supported platforms

.. image:: https://readthedocs.org/projects/pylexique/badge/?version=latest
        :target: https://pylexique.readthedocs.io/en/latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/SekouDiaoNlp/pylexique/shield.svg
        :target: https://pyup.io/repos/github/SekouDiaoNlp/pylexique/
        :alt: Dependencies status

.. image:: https://codecov.io/gh/SekouDiaoNlp/pylexique/branch/master/graph/badge.svg?token=EiEXyUJGpF
        :target: https://codecov.io/gh/SekouDiaoNlp/pylexique
        :alt: Code Coverage Status

.. image:: https://snyk-widget.herokuapp.com/badge/pip/pylexique/badge.svg
        :target: https://snyk.io/test/github/SekouDiaoNlp/pylexique?targetFile=requirements.txt
        :alt: Code Vulnerability Status

.. image:: https://img.shields.io/conda/dn/conda-forge/pylexique?label=Anaconda%20Total%20Downloads
        :target: https://anaconda.org/conda-forge/pylexique
        :alt: Conda


|


| Pylexique is a Python wrapper around Lexique383_.

| It allows the extraction of lexical information from more than 140 000 French words in an Object Oriented way.

|
|

* Free software: MIT license
* PyLexique Documentation: https://pylexique.readthedocs.io (en) -- https://sekoudiaonlp.github.io/pylexique/fr_FR/ (fr)

| Each lexical item is represented as a LexItem having the following LexEntryType:
|


.. code-block:: python

        class LexEntryType:
        """
        Type information about all the lexical attributes in a LexItem object.

        """
        ortho: str
        phon: str
        lemme: str
        cgram: str
        genre: str
        nombre: str
        freqlemfilms2: float
        freqlemlivres: float
        freqfilms2: float
        freqlivres: float
        infover: str
        nbhomogr: int
        nbhomoph: int
        islem: bool
        nblettres: int
        nbphons: int
        cvcv: str
        p_cvcv: str
        voisorth: int
        voisphon: int
        puorth: int
        puphon: int
        syll: str
        nbsyll: int
        cv_cv: str
        orthrenv: str
        phonrenv: str
        orthosyll: str
        cgramortho: str
        deflem: float
        defobs: int
        old20: float
        pld20: float
        morphoder: str
        nbmorph: int

The meanings of the attributes of this object are as follow:

* ortho: the word
* phon: the phonological forms of the word
* lemme: the lemmas of this word
* cgram: the grammatical categories of this word
* genre: the gender
* nombre: the number
* freqlemfilms2: the frequency of the lemma according to the corpus of subtitles (per million occurrences)
* freqlemlivres: the frequency of the lemma according to the body of books (per million occurrences)
* freqfilms2: the frequency of the word according to the corpus of subtitles (per million occurrences)
* freqlivres: the frequency of the word according to the body of books (per million occurrences)
* infover: modes, tenses, and possible people for verbs
* nbhomogr: number of homographs
* nbhomoph: number of homophones
* islem: indicates if it is a lemma or not
* nblettres: the number of letters
* nbphons: number of phonemes
* cvcv: the orthographic structure
* p-cvcv: the phonological structure
* voisorth: number of orthographic neighbors
* voisphon: number of phonological neighbors
* puorth: point of spelling uniqueness
* puphon: point of phonological uniqueness
* syll: syllable phonological form
* nbsyll: number of syllables
* cv-cv: syllable phonological structure
* orthrenv: reverse orthographic form
* phonrenv: reversed phonological form
* orthosyll: syllable orthographic form
* cgramortho: the different grammatical category for a given orthographic representation
* deflem: the percentage of people who said they knew the lemma of the word
* defobs: the size of the sample from which 'deflem' is derived
* old20:  orthographic Levenshtein Distance
* pld20: phonological Levenshtein Distance
* morphoder: inflectional morphology
* nbmorph: the number of morphemes directly computed from 'morphoder'


You can find all the relevant information in the `official documentation of Lexique383`_ (French).


Features
--------

* Extract all lexical information from a French  word such as:
    * orthographic and phonemics representations
    * associated lemmas
    * syllabation
    * grammatical category
    * gender and number
    * frequencies in a corpus of books and in a body of film subtitles, etc...
* Extract all the lexical forms of a French word.
* Easy to use Api.
* Easily integrate pylexique in your own projects as an imported library.
* Can be used as a command line tool.

Credits
-------

Main developer SekouDiaoNlp_.

Lexical corpus: Lexique383_

About Lexique383
================

Lexique3
========

Lexique 3.83 is a French lexical database that provides
for ~ 140,000 words of French: orthographic and phonemics representations,
associated lemmas, syllabation, grammatical category, gender and number,
frequencies in a corpus of books and in a body of film subtitles, etc...

|

Table: `Lexique383.zip`_

Web site: http://www.lexique.org

Online: http://www.lexique.org/shiny/lexique

Publications
------------

-  New, Boris, Christophe Pallier, Marc Brysbaert, and Ludovic Ferrand.
   2004. "Lexique 2: A New French Lexical Database." *Behavior Research
   Methods, Instruments, & Computers* 36 (3): 516--524.
   `DOI <https://doi.org/10.3758/bf03195598>`__.
   `pdf`_

-  New, Boris, Christophe Pallier, Ludovic Ferrand, and Rafael Matos.
   2001. "Une Base de Données Lexicales Du Français Contemporain Sur
   Internet: LEXIQUE" *L'Année Psychologique* 101 (3): 447--462.
   `DOI <https://doi.org/10.1017/S014271640707035X>`__.
   `pdf <https://docs.google.com/viewer?url=http://www.lexique.org/outils/Lexique_Annee.pdf>`__

-  Boris New, Marc Brysbaert, Jean Veronis, and Christophe Pallier.
   2007. "The Use of Film Subtitles to Estimate Word Frequencies."
   Applied Psycholinguistics 28 (4): 661--77.
   `DOI <https://doi.org/10.1017/S014271640707035X>`__.
   (`pdf <https://drive.google.com/file/d/1uvKrxGqETXkFeRH4PcYaql8ES9FjEdbV/view?usp=sharing>`__)

Contributors
------------

-  Boris New & Christophe Pallier
-  Ronald Peereman
-  Sophie Dufour
-  Christian Lachaud
-  and many others... (contact us to be listed)

License
-------

`CC BY SA40.0`_

.. _Lexique383.zip: http://www.lexique.org/databases/Lexique383/Lexique383.zip
.. _pdf: https://docs.google.com/viewer?url=http://sites.google.com/site/borisnew/pub/New-et-al2004-BRMIC.pdf?attredirects=0
.. _CC BY SA40.0: LICENSE-CC-BY-SA4.0.txt


BibTex Entry to cite publications about Lexique383:


.. code:: bibtex

    @article{npbf04,
    author = {New, B. and Pallier, C. and Brysbaert, M. and Ferrand, L.},
    journal = {ehavior Research Methods, Instruments, & Computers},
    number = {3},
    pages = {516-524},
    title = {Lexique 2 : A New French Lexical Database},
    volume = {36},
    year = {2004},
    eprint = {http://www.lexique.org/?page_id=294},
    }

.. code:: bibtex

    @article{npfm01,
    author = {New, B. and Pallier, C. and Ferrand, L. and Matos, R.},
    journal = {L'Ann{\'e}e Pschologique},
    number = {447-462},
    pages = {1396-2},
    title = {Une base de donn{\'e}es lexicales du fran\c{c}ais contemporain sur internet: LEXIQUE},
    volume = {101},
    year = {2001},
    }

.. code:: bibtex

    @article{new_brysbaert_veronis_pallier_2007,
    author={NEW, BORIS and BRYSBAERT, MARC and VERONIS, JEAN and PALLIER, CHRISTOPHE},
    title={The use of film subtitles to estimate word frequencies},
    volume={28}, DOI={10.1017/S014271640707035X},
    number={4}, journal={Applied Psycholinguistics},
    publisher={Cambridge University Press},
    year={2007},
    pages={661–677}}

BibTeX
------

If you want to cite pylexique in an academic publication use this citation format:

.. code:: bibtex

   @article{pylexique,
     title={pylexique},
     author={Sekou Diao},
     journal={GitHub. Note: https://github.com/SekouDiaoNlp/pylexique Cited by},
     year={2021}
   }


.. _Lexique383: http://www.lexique.org/
.. _SekouDiaoNlp: https://github.com/SekouDiaoNlp
.. _`official documentation of Lexique383`: http://lexique.org/_documentation/Manuel_Lexique.3.2.pdf


HISTORY.rst
=======
History
=======

1.5.1 (2023-08-23)
------------------

* Fixed issue with certain words not being displayed properly.
* Updated dependencies.

1.5.0 (2021-10-28)
------------------

* Added new method 'Lexique383.get_anagrams()' to get the list of all the anagrams of a given word.
* Removed support for xlsb file format..
* Migrated the project build configuration to pyproject.toml.
* Migrated build backend to poetry.
* Updated the Documentation and the docstrings.
* Updated dependencies.

1.4.0 (2021-10-23)
------------------

* The library is now completely type annotated and type-checked.
* The method get_all_forms() now checks if the word has multiple lemmas in case of homonymy, eg: 'souris'.
* Fixed parsing bug when using 'std-csv' parser.
* Fixed formatting issue when saving the output of the CLI to a file.
* Updated the Documentation and the docstrings.
* Updated dependencies.

1.3.5 (2021-05-18)
------------------

* Uses str.lower() to normalize inputs.
* Updated the Documentation and the docstrings.

1.3.4 (2021-05-16)
------------------

* Fixed bug where Lexique383 was not shipped with the distribution.
* Made csv parsing far faster and more robust.
* Can now use different parsers : 'pandas_csv' is the pandas csv parser, 'std_csv' is the standard library csv parser, 'csv' is a custom csv parser and 'xlsb' is pandas xlsb parser using pyxlsb engine.
* Updated dependencies.

1.3.3 (2021-05-16)
------------------

* Made csv parsing faster and more robust.
* Can now use different parsers : 'pandas_csv' is the pandas csv parser, 'std_csv' is the standard library csv parser, 'csv' is a custom csv parser and 'xlsb' is pandas xlsb parser using pyxlsb engine.
* Updated dependencies.

1.3.2 (2021-05-14)
------------------

* Can now use both 'csv' and 'xlsb' files.
* Uses 'csv' file for storage and faster load times.
* Updated dependencies.

1.3.1 (2021-05-12)
------------------

* Uses pandas for now for faster resource loading.
* Uses xlsb file for storage and faster load times
* Updated dependencies.

1.3.0 (2021-05-11)
------------------

* Uses pandas for now for faster resource loading.
* In the process of integrating `faster-than-csv` when MacOs issues get resolved.
* Refactored and expanded the test suite.
* Updated dependencies.

1.2.7 (2021-05-07)
------------------

* The new method Lexique383.get_all_forms(word) is now accessible through the cli with option '-a' or '--all_forms'.
* This new method returns a list of LexItems having the same root lemma.
* Added sample commands using the new option in the docs.
* Refactored and expanded the test suite.
* Updated dependencies.

1.2.6 (2021-05-06)
------------------

* allows for new style of relative imports.
* Now all the attributes of the LexItem objects are immutable for consistency.
* Added new method Lexique383.get_all_forms(word) to get all the lexical variations of a word.
* This new method returns a list of LexItems having the same root lemma.
* Expanded sample usage of the software in the docs.
* Updated dependencies.

1.2.3 (2021-05-04)
------------------

* Enhanced behaviour of output to stdout to not conflict with the logging strategy of users importing the library in their own projects.
* Expanded sample usage of the software in the docs.
* Updated dependencies.

1.2.2 (2021-05-04)
------------------

* Enhanced Type Hinting for main module.
* Changed the property LexItem.islem to boolean instead of a binary choice 0/1.
* Expanded sample usage of the software in the docs.
* Updated dependencies.

1.2.1 (2021-04-30)
------------------

* Implemented Type Hinting for main module.
* Added a new method to the class Lexique383. The method is Lexique383._save_errors() .
* This new method checks that the value of each field in a LexItem is of the right type. If it finds errors it will record the mismatched value/type and save it in ./erros/errors.json
* Expanded sample usage of the software in the docs.
* Much better documentation including links to Lexique383 pages and manuals.

1.2.0 (2021-04-30)
------------------

* Added a new method to the class Lexique383. The method is Lexique383.get_lex() .
* This new method accepts either a single word as a string or an iterable of strings and will return the asked lexical information.
* Expanded sample usage of the software in the docs.
* Substantial update to the code and docs.
* Removed unneeded dependencies as I reimplement some functionality myself.

1.1.1 (2021-04-28)
------------------

* Added a new method to the class LexItem. The method is LexItem.to_dict() .
* This new method allows the LexItem objects to be converted into dicts with key/value pairs corresponding to the LexItem.
* This method allows easy display or serialization of the LexItem objects.
* Lexical Items having the same orthography are stored in a list at the word's orthography key to the LEXIQUE dict.
* Expanded sample usage of the software in the docs.
* Substantial update to the code and docs.

1.1.0 (2021-04-28)
------------------

* Drastically reduced dependencies by ditching HDF5 and bolcs as the package is now smaller, faster an easier to build.
* Lexical Items having the same orthography are stored in a list at the word's orthography key to the LEXIQUE dict.
* Implemented the "FlyWheel" pattern for light Lexical entries rsiding entirely in memory at run time.
* Added sample usage of the software in the docs.
* General update to the code and docs.

1.0.7 (2021-04-27)
------------------

* First release on PyPI.


pylexique/cli.py
# -*- coding: utf-8 -*-

"""Console CLI script for pylexique."""
import sys
import click
import json
import logging
from rich.console import Console
from rich.table import Table
from pylexique import Lexique383, LexItem
from collections import defaultdict
from typing import Sequence, Dict, Union, List, DefaultDict

LEXIQUE383_FIELD_NAMES = ['ortho', 'phon', 'lemme', 'cgram', 'genre', 'nombre', 'freqlemfilms2', 'freqlemlivres',
                          'freqfilms2',
                          'freqlivres', 'infover', 'nbhomogr', 'nbhomoph', 'islem', 'nblettres', 'nbphons', 'cvcv',
                          'p_cvcv',
                          'voisorth', 'voisphon', 'puorth', 'puphon', 'syll', 'nbsyll', 'cv_cv', 'orthrenv', 'phonrenv',
                          'orthosyll', 'cgramortho', 'deflem', 'defobs', 'old20', 'pld20', 'morphoder', 'nbmorph', 'sorted_ortho']
def convert_to_dict(obj: LexItem) -> Dict[str, Union[str, float, int, bool]]:
    if isinstance(obj, LexItem):
        return obj.to_dict()
    return obj

def _display_results(console: Console, results: DefaultDict[str, List[Union[LexItem, List[LexItem]]]], output: str) -> None:
    """Display lexical results using rich tables."""
    for word, elements in results.items():
        if not elements:
            console.print(f"The word {word} was not found in Lexique383.")
            continue
        table = Table(title=f"Lexical Information for '{word}'", show_header=True)
        table.add_column("Attribute", style="bold")

        num_columns = 1  # Initialize the number of columns
        for element in elements:
            if isinstance(element, LexItem):
                num_columns += 1
                table.add_column(element.lemme, justify="center")  # Add a column for each LexItem
                element_dict = element.to_dict()
                for field in LEXIQUE383_FIELD_NAMES:
                    value = element_dict.get(field, "")
                    table.add_row(field, str(value), *[None] * (num_columns - 2))  # Add None values for other columns
            else:
                for item in element:
                    if isinstance(item, LexItem):
                        num_columns += 1
                        table.add_column(item.lemme, justify="center")  # Add a column for each LexItem
                        item_dict = item.to_dict()
                        for field in LEXIQUE383_FIELD_NAMES:
                            value = item_dict.get(field, "")
                            table.add_row(field, *[None] * (num_columns - 2), str(value))  # Add None values for other columns

        console.print(table)

    if output:
        with open(output, 'w', encoding='utf-8') as file:
            json.dump(results, file, indent=4, ensure_ascii=False, default=convert_to_dict)
            console.print(f"The Lexical Items have been successfully saved to {output} by pylexique.")

@click.command(context_settings=dict(help_option_names=["-h", "--help"])) #type: ignore
@click.argument('words', nargs=-1) #type: ignore
@click.option('-a', '--all_forms', #type: ignore
              is_flag=True,
              help="Gets all lexical forms of a given word. Only takes 1 word as an argument.")
@click.option('-o', '--output', #type: ignore
              default=None,
              help="Path of the json filename for storing the lexical entries.",
              type=click.STRING)
@click.option('-i', '--interactive', is_flag=True, help="Enter interactive mode to input words and options interactively.") #type: ignore
def main(words: Sequence[str], all_forms: bool, output: str, interactive: bool) -> None:
    """Pylexique is a Python wrapper around Lexique83.
    It allows to extract lexical information from more than 140 000 French words in an Object Oriented way.


    * Free software: MIT license
    * Documentation: https://pylexique.readthedocs.io.
    """
    logger = logging.getLogger(__name__)

    # create console handler and set level to debug
    console_handler = logging.StreamHandler(sys.stdout)
    error_handler = logging.StreamHandler(sys.stderr)
    console_handler.setLevel(logging.INFO)
    error_handler.setLevel(logging.ERROR)
    logger.addHandler(console_handler)
    logger.addHandler(error_handler)
    logger.setLevel(logging.INFO)

    LEXIQUE = Lexique383()
    
    if interactive:
        _run_interactive_mode(LEXIQUE, output)
    else:
        _run_batch_mode(LEXIQUE, words, all_forms, output)

def _run_interactive_mode(lexique: Lexique383, output: str) -> None:
    """Run the interactive mode for pylexique."""
    console = Console()
    cached_results: Dict[str, DefaultDict[str, List[Union[LexItem, List[LexItem]]]]] = {}

    while True:
        word = click.prompt("Enter a word (or press Ctrl+C to quit):", type=str)
        
        try:
            if word in cached_results:
                results = cached_results[word]
            else:
                all_forms = click.confirm("Get all forms of the word?")
                results = _get_results(lexique, [word], all_forms)
                cached_results[word] = results

            _display_results(console, results, output)
        except KeyboardInterrupt:
            console.print("\nExiting interactive mode.")
            break

def _run_batch_mode(lexique: Lexique383, words: Sequence[str], all_forms: bool, output: str) -> None:
    """Run the batch mode for pylexique."""
    console = Console()
    results = _get_results(lexique, words, all_forms)
    _display_results(console, results, output)

def _get_results(lexique: Lexique383, words: Sequence[str], all_forms: bool) -> DefaultDict[str, List[Union[LexItem, List[LexItem]]]]:
    """Get lexical results for the provided words."""
    results: DefaultDict[str, List[Union[LexItem, List[LexItem]]]] = DefaultDict(list)
    for word in words:
        lex_items = lexique.get_lex(word).get(word, [])
        if lex_items:
            if all_forms:
                results[word].append(lexique.get_all_forms(word))
            else:
                results[word].append(lex_items)
        else:
            results[word] = []
    return results

if __name__ == "__main__":
    main()  # pragma: no cover
